examples/usdt_sample/scripts/latency.py

```diff
@@ -0,0 +1,143 @@
+};
+
+/**
+ * @brief Contains the latency data w.r.t. the complete operation from request to response.
+ */
+struct end_data_t
+{
+    u64 operation_id; ///< The id of the operation.
+    char input[64];   ///< The request (input) string.
+    char output[64];  ///< The response (output) string.
+    u64 start;        ///< The start timestamp of the operation.
+    u64 end;          ///< The end timestamp of the operation.
+    u64 duration;     ///< The duration of the operation.
+};
+
+/**
+ * @brief Contains the operation start data.
+ * key: the operation id.
+ * value: The operation start latency data.
+ */
+BPF_HASH(start_hash, u64, struct start_data_t);
+
+/**
+ * The output buffer, which will be used to push the latency event data to user space.
+ */
+BPF_PERF_OUTPUT(operation_event);
+
+/**
+ * @brief Reads the operation request arguments and stores the latency data in the hash.
+ * @param ctx The BPF context.
+ */
+int trace_operation_start(struct pt_regs* ctx)
+{
+    struct start_data_t start_data = {};
+    bpf_usdt_readarg(1, ctx, &start_data.operation_id);
+    bpf_usdt_readarg_p(2, ctx, &start_data.input, sizeof(start_data.input));
+    start_data.start = bpf_ktime_get_ns();
+    start_hash.update(&start_data.operation_id, &start_data);
+    return 0;
+}
+
+/**
+ * @brief Reads the operation response arguments, calculates the latency event data, and writes it to the user output buffer.
+ * @param ctx The BPF context.
+ */
+int trace_operation_end(struct pt_regs* ctx)
+{
+    struct end_data_t end_data = {};
+    bpf_usdt_readarg(1, ctx, &end_data.operation_id);
+    bpf_usdt_readarg_p(2, ctx, &end_data.output, sizeof(end_data.output));
+    end_data.end = bpf_ktime_get_ns();
+
+    struct start_data_t* start_data = start_hash.lookup(&end_data.operation_id);
+    if (0 != start_data) {
+        end_data.start = start_data->start;
+        end_data.duration = end_data.end - end_data.start;
+        bpf_probe_read(end_data.input, sizeof(end_data.input), start_data->input);
+    }
+
+    operation_event.perf_submit(ctx, &end_data, sizeof(end_data));
+    start_hash.delete(&end_data.operation_id);
+    return 0;
+}
+"""
+
+# Create USDT context
+print("Attaching probes to pid %d" % this_pid)
+usdt_ctx = USDT(pid=this_pid)
+usdt_ctx.enable_probe(probe="operation_start", fn_name="trace_operation_start")
+usdt_ctx.enable_probe(probe="operation_end", fn_name="trace_operation_end")
+
+# Create BPF context, load BPF program
+bpf_ctx = BPF(text=latency_bpf_text, usdt_contexts=[usdt_ctx], debug=debugLevel)
+
+# Potential issue: when using a separate file (i.e. latency.c), the following error occurs:
+# LLVM ERROR: Program used external function '_bpf_readarg_trace_operation_end_1' which could not be resolved!
+#latency_src_file = "%s/latency.c" % os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe())))
+#bpf_ctx = BPF(src_file=latency_src_file, usdt_contexts=[usdt_ctx], debug=debugLevel)
```

Comment:
If there are code duplication, you can still put into a common C file and read it into a string in python code. If there are minor differences between different C files, you can use macro's to differentiate them and use -D to control each individual case.